src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -98,104 +235,213 @@ open class GossipRouter : AbstractRouter() {
                     }
+                    .flatten()
             }
-            .flatten()
-            .map { submitPublishMessage(it, msg) }
+        val list = peers.map { submitPublishMessage(it, msg) }
 
-        mCache.put(msg)
+        mCache.put(getMessageId(msg), msg)
         flushAllPending()
         return anyComplete(list)
     }
 
-    override fun subscribe(topic: String) {
+    override fun subscribe(topic: Topic) {
         super.subscribe(topic)
-        val fanoutPeers = fanout[topic] ?: mutableSetOf()
+        val fanoutPeers = (fanout[topic] ?: mutableSetOf())
+            .filter { score.score(it) >= 0 && !isDirect(it) }
         val meshPeers = mesh.getOrPut(topic) { mutableSetOf() }
-        val otherPeers = getTopicPeers(topic) - meshPeers - fanoutPeers
-        if (meshPeers.size < D) {
-            val addFromFanout = fanoutPeers.shuffled(random).take(D - meshPeers.size)
-            val addFromOthers = otherPeers.shuffled(random).take(D - meshPeers.size - addFromFanout.size)
+        val otherPeers = (getTopicPeers(topic) - meshPeers - fanoutPeers)
+            .filter { score.score(it) >= 0 && !isDirect(it) }
+
+        if (meshPeers.size < coreParams.D) {
+            val addFromFanout = fanoutPeers.shuffled(random)
+                .take(coreParams.D - meshPeers.size)
+            val addFromOthers = otherPeers.shuffled(random)
+                .take(coreParams.D - meshPeers.size - addFromFanout.size)
 
-            meshPeers += (addFromFanout + addFromOthers)
             (addFromFanout + addFromOthers).forEach {
                 graft(it, topic)
             }
+            fanout -= topic
+            lastPublished -= topic
         }
     }
 
-    override fun unsubscribe(topic: String) {
+    override fun unsubscribe(topic: Topic) {
         super.unsubscribe(topic)
-        mesh.remove(topic)?.forEach { prune(it, topic) }
+        mesh[topic]?.forEach { prune(it, topic) }
     }
 
-    private fun heartBeat(time: Long) {
+    private fun heartBeat() {
+        heartbeatsCount++
+        iAsked.clear()
+        peerIHave.clear()
+
+        val staleIWantTime = this.curTime() - params.iWantFollowupTime.toMillis()
+        iWantRequests.entries.removeIf { (key, time) ->
+            (time < staleIWantTime)
+                .whenTrue { notifyRouterMisbehavior(key.first, 1) }
+        }
+
         try {
             mesh.entries.forEach { (topic, peers) ->
-                if (peers.size < DLow) {
-                    (getTopicPeers(topic) - peers).shuffled(random).take(D - peers.size).forEach { newPeer ->
-                        peers += newPeer
-                        graft(newPeer, topic)
-                    }
-                } else if (peers.size > DHigh) {
-                    peers.shuffled(random).take(peers.size - D).forEach { dropPeer ->
-                        peers -= dropPeer
-                        prune(dropPeer, topic)
+
+                // drop underscored peers from mesh
```

Comment:
Should we not be doing the same thing to fanout topic and peers?

Reply:
Fanout peers are not in the mesh and we can't prune them, however we are also cleaning our fanout list from underscored peers: 
https://github.com/libp2p/jvm-libp2p/blob/7fc89c886a04fddafc123c3dc14d19592cbd92ba/src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt#L365